### PR TITLE
OKTA-488781 Password enroll/expire/reset with confirmation component implementation

### PR DESCRIPTION
**Pull request by glenfannin-okta**
## Description:
Purpose of this PR is to implement the custom Password with confirmation component.

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [x] Added e2e tests

- [x] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

- Did you verify the change by running downstream monolith artifacts? (YES | NO | UNSURE)

### Screenshot/Video:

https://user-images.githubusercontent.com/97472729/182455720-2caebfc4-e9f3-4631-b14c-4d7ce2b52540.mov

### Issue:

- [OKTA-488781](https://oktainc.atlassian.net/browse/OKTA-488781)

### CODE REVIEW

File: src/v3/src/util/resetMessagesToInputs.ts
Comment (shuowu-okta) [inline src/v3/src/util/resetMessagesToInputs.ts:17]: renamed due to naming collision?

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (shuowu-okta) [inline src/v3/test/integration/authenticator-expired-password.test.tsx:79]: be specific this field message comes from confirmPassword field

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (shuowu-okta) [inline src/v3/test/integration/authenticator-expired-password.test.tsx:101]: be specific this field message comes from new password field

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (shuowu-okta) [inline src/v3/test/integration/authenticator-expired-password.test.tsx:127]: be specific this field message comes from confirmPassword field

File: src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts
Comment (shuowu-okta) [inline src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts:66]: how about following the `inputMeta` pattern for FieldElement?

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:88]: split this handler function into two separate ones and attach each to the corresponding field

File: src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts
Comment (shawyu-okta) [inline src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts:64]: not a required change: but wondering if we ever have an instance where we need the flexibility to pass the label for this element type. if not, we can consider encapsulating this label

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:56]: can we focus this PR on form submission validation only, and handle this state in the following PR?

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shawyu-okta) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:117]: related to my other comment about the label being encapsulated -- if it were encapsulated it would mean we could avoid this type assertion as well

**Review by shawyu-okta — APPROVED** [review]
minor comments

**Comment by shawyu-okta** [comment]
I noticed this case in the video, is this expected? Seems weird that it reports the first password field as blank when it isn't (I'd expect it to say passwords don't match or only say the field is blank for the second input)
image

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:116]: You can have two uischema states, one for password, and one for confirmPassword, and pass each uischema to its InputPassword component. 

Also make InputPassword can accept `value` props to override `const value = useValue(uischema)`, so you can pass state down to that component.

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:98]: can be two functions, and let's handle it in the following PR, so the feature comes in with tests

**Review by shuowu-okta — COMMENTED** [review]
Thanks for separating this change out!

**Comment by glenfannin-okta** [comment]
> I noticed this case in the video, is this expected? Seems weird that it reports the first password field as blank when it isn't (I'd expect it to say passwords don't match or only say the field is blank for the second input) image

This is expected in the current state of the widget, I reported this issue here: https://oktainc.atlassian.net/browse/OKTA-518428 it is caused by this because when submitting, it clears the data bag, but doesn't reflect that change in our inputs.

File: src/v3/src/util/resetMessagesToInputs.ts
Comment (glenfannin-okta) [reply src/v3/src/util/resetMessagesToInputs.ts:17]: Yes and no, this prop looks like this 
```
{ 
   "identifier": [{ i18n: { key: 'some.key'}}],
   "credentials.passcode": [{ i18n: { key: 'some.key'}}]
}
```
The name of the prop being "messages" isnt exactly reflective of what it stores. This is one reason why I renamed it, the second reason is because we actually grab the messages from the record below, and I wanted to use that variable name.

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (glenfannin-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx:79]: in this case, there can only be one error message, if there are multiple with this text jest will throw an error.

Comment (shuowu-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx:79]: but it's not clear which field this error message belongs to, right?

File: src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts
Comment (glenfannin-okta) [reply src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts:64]: Do you mean pass the `label` as a prop to `options` instead of here? 

I can do that, I just re-used the `label` attr that was available for the confirm password field to prevent duplicating props in the schema.

Comment (shawyu-okta) [reply src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts:64]: I mean if `label` can live hardcoded inside the `PasswordWithConfirmation` component

Comment (glenfannin-okta) [reply src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts:64]: Ohhh, the direction that was told to me was that the components should not use `loc` or any translation. That the components should always receive translated text.

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (shawyu-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx:79]: agreed @shuowu-okta I think it would be good to scope this `find*` selector within the specific input we care about rather than the global `screen` selectors

**Comment by shawyu-okta** [comment]
> This is expected in the current state of the widget, I reported this issue here: https://oktainc.atlassian.net/browse/OKTA-518428 it is caused by this because when submitting, it clears the data bag, but doesn't reflect that change in our inputs.

thanks for the confirming!

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (glenfannin-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx:79]: The test description does, but I will update. :)

Comment (shuowu-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx:79]: It's not the same, think about if this error message appears in new password field, test will still pass, but that will be a bug in widget

Comment (glenfannin-okta) [reply src/v3/test/integration/authenticator-expired-password.test.tsx]: Yep, forgot this is an integration test and should guard against other issues like this. Updated.

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (glenfannin-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:51]: The other Pr is already open (and I am waiting for this to merge to accept and resolve conflicts), I think you are looking to have a HOC somewhat contain this type of state which I can continue looking into after this gets merged and resolve them together. 

But this is to at least match parity for now and drive this feature to completion and refactor (if needed) in the other PR that is open.

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (glenfannin-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:103]: Other than the label and name, what benefit do you get out of creating a separate schema element that is not from the transaction, that will only be used in this component?

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:51]: this will end up with not tested code get merged in, let's be focus in this PR for submission validation only

File: src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts
Comment (glenfannin-okta) [reply src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts]: I made this update, but, wouldnt it be confusing to someone that looks at `inputMeta` in the `FieldElement` and see its signature, then come here and expect that same signature but to realize it is actually a `FieldElement`?

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (glenfannin-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:51]: I can add a regression test to prevent this.

Also, will stage and commit my unit test.

File: src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts
Comment (shuowu-okta) [reply src/v3/src/transformer/password/transformEnrollPasswordAuthenticator.ts]: this one might be a special case,  maybe we can add `password` and `confirmPassword` to contain uischemas for both fields, or you can derive internal uischema states. Not sure which one works better

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:51]: then what's the purpose of splitting the changes, it's getting together into another mega PR

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:103]: DRY

Comment (shuowu-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:103]: if the component is hard to be reused, which most probably means that component is not well organized

Comment (glenfannin-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:103]: Are you saying to do something like this? 

```
export interface PasswordWithConfirmationElement extends UISchemaElement {
  type: 'PasswordWithConfirmation';
  options: {
    password: FieldElement,
    confirmPassword: FieldElement,
  };
}
```

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (glenfannin-okta) [reply src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:51]: I will remove it and handle it in the next PR, but i will create a test and skip it to make sure it gets covered.

File: src/v3/test/integration/authenticator-expired-password.test.tsx
Comment (glenfannin-okta) [inline src/v3/test/integration/authenticator-expired-password.test.tsx:143]: An odd issue here is happening (only in test) after rebasing with 0.0, the validate function is not being overwritten. When I manually test, it works just fine and I can see both errors, but in this test it fails because the function isnt overwritten. I debugged through it and I can see the validate function being re-set but still doesnt get executed. Still debugging, but wanted to push up the other changes for re-review in the meantime.

File: src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx
Comment (shuowu) [inline src/v3/src/components/PasswordWithConfirmation/PasswordWithConfirmation.tsx:60]: nit: you can also InputPasswprd to accept value and onChange props to override the default behavior, then it can be reused here

File: src/v3/src/components/Widget/index.tsx
Comment (shuowu) [inline src/v3/src/components/Widget/index.tsx:79]: you can add comment here, like data not for form submission

**Review by shuowu — APPROVED** [review]

**Pull request merged**
